gateway/api_loader.go

```diff
@@ -692,6 +692,12 @@ func loadApps(specs []*APISpec) {
 
 	// Swap in the new register
 	apisMu.Lock()
+
+	// release current specs resources before overwriting map
+	for _, curSpec := range apisByID {
+		curSpec.Release()
+	}
+
 	apisByID = tmpSpecRegister
 	apisMu.Unlock()
```

Comment:
I think the locking logic here can be improved. 

```go
	apisMu.RLock()
	// release current specs resources before overwriting map
	for _, curSpec := range apisByID {
		curSpec.Release()
	}
	apisMu.RUnlock()

	apisMu.Lock()
	apisByID = tmpSpecRegister
	apisMu.Unlock()
```

We are first reading from `apisByID` so we use `RLock`  and then writing to it using `Lock`

Reply:
In this implementation, it should be fine.